flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(classpathProvider.getJobClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromSystemClasspathExtractingTheJobClassFromThere()
+            throws IOException, FlinkException {
+        classpathProvider.setSystemClasspathWithEntryClass();
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, null, null, null, new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(classpathProvider.getJobClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromClasspathWithJobClass()
+            throws IOException, FlinkException {
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        // the user directory must be specified
+                        classpathProvider.getUserDirectoryWithTwoEntryClasses(),
+                        // the user classpath is derived from the user directory outside of the
+                        // method
+                        classpathProvider.getURLUserClasspathWithTwoEntryClasses(),
+                        null,
+                        // we have to specify the job class - otherwise the call would fail due to
+                        // two main method being present
+                        classpathProvider.getJobClassName(),
+                        new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(classpathProvider.getJobClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationFromClasspathWithNoJobClass()
+            throws IOException, FlinkException {
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        // the user directory must be specified
+                        classpathProvider.getUserDirectoryWithOneEntryClass(),
+                        // the user classpath is derived from the user directory outside of the
+                        // method
+                        classpathProvider.getURLUserClasspathWithEntryClass(),
```

Comment:
which method are you referencing?

Reply:
The method that's being tested. I updated the comment to make this more clear